File: api/github.go
package api

import (
	"net/http"

	"github.com/labstack/echo/v4"
	"isabelroses.com/lib"
)

func Github(c echo.Context) error {
	g := new(lib.GitHubDono)

	if err := c.Bind(g); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, err.Error())
	}

	if g.Sponsorship.PrivaryLevel == "SECRET" {
		return c.String(http.StatusOK, "Accepted, and hidden")
	}

	if g.Action != "created" {
		return c.String(http.StatusOK, "Not a new sponsorship")
	}

	var newData map[string]interface{}
	switch {
	case g.Sponsorship.Tier.IsOneTime:
		newData = map[string]interface{}{
			"tier":   "OneTime",
			"name":   getName(g.Sponsorship.Sponsor),
			"avatar": g.Sponsorship.Sponsor.AvatarURL,
		}
	case !g.Sponsorship.Tier.IsOneTime:
		newData = map[string]interface{}{
			"tier":   g.Sponsorship.Tier.Name,
			"name":   getName(g.Sponsorship.Sponsor),
			"avatar": g.Sponsorship.Sponsor.AvatarURL,
		}
	}

	lib.AppendTooDonos(newData)

	return c.JSON(http.StatusOK, g)
}

func getName(sponsor lib.GitHubSponsor) string {
	if sponsor.Name != "" {
		return sponsor.Name
	}
	return sponsor.Login
}

File: src/api/github.rs
use axum::{http::StatusCode, response::IntoResponse, Json};
use serde::{Deserialize, Serialize};

use crate::data::donos;

#[derive(Debug, Deserialize, Serialize)]
pub struct Dono {
    action: String,
    sponsorship: Sponsorship,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Sponsorship {
    created_at: String,
    privacy_level: String,
    sponsor: Sponsor,
    tier: Tier,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Sponsor {
    avatar_url: Option<String>,
    login: String,
    html_url: Option<String>,
    name: Option<String>,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct Tier {
    name: String,
    is_one_time: bool,
}

pub async fn handler(Json(payload): Json<Dono>) -> impl IntoResponse {
    if payload.sponsorship.privacy_level == "SECRET" {
        return (StatusCode::OK, "Accepted, and hidden".to_string()).into_response();
    }

    // Only process newly created sponsorships
    if payload.action != "created" {
        return (StatusCode::OK, "Not a new sponsorship".to_string()).into_response();
    }

    let new_data = donos::Donor {
        tier: if payload.sponsorship.tier.is_one_time {
            "OneTime".to_string()
        } else {
            payload.sponsorship.tier.name.clone()
        },
        name: get_name(&payload.sponsorship.sponsor),
        url: payload
            .sponsorship
            .sponsor
            .html_url
            .unwrap_or(String::new()),
        avatar: payload
            .sponsorship
            .sponsor
            .avatar_url
            .unwrap_or(String::new()),
    };

    donos::add(new_data.clone());

    (StatusCode::OK, Json(new_data)).into_response()
}

fn get_name(sponsor: &Sponsor) -> String {
    sponsor
        .name
        .clone()
        .unwrap_or_else(|| sponsor.login.clone())
}
